### DISCUSSION TITLE
Model doesn't learn

### DISCUSSION THREAD

**Discussion by gkrisp98** [TorchIO-project/torchio#801] [Ideas]
Hi, I am new to the field of medical images. I am working with 435 MRIs and I have to classify them to 5 classes. I have created a subject dataset both for nifti and DICOM files but neither of them is learning.  The training and validation accuracy range from 20 to 22% which is just random. 
I used the transforms RandomAffine(), RescaleIntensity(), RandomNoise and CropOrPad() to make the shape of the objects (160,224,224) as it was (160,384,384).
As of now I used a efficient net 3D, a resnet3d and a resnet18 from monai but neither of them seems to learn anything.
Is there any advice for what I can do to make my models start learning ?

**Comment by fepegar** [comment]
Hi, @gkrisp98. There are a thousand potential causes. Take a look at Andrej Karpathy's [recipe](http://karpathy.github.io/2019/04/25/recipe/).

If you're using MONAI's tools, I suggest you ask on their GitHub, as TorchIO might be unrelated to the issue.

**Reply by gkrisp98** [reply]
Thanks a lot! I will check it.

**Reply by gkrisp98** [reply]
I have another question if you can help me. Is it ok to use CropOrPard() to change the shape of my image ? I know that I am loosing information but (160,384,384) is too big and CUDA runs out of memory even with batch_size = 1

**Reply by fepegar** [reply]
Sure. You can also use [patch-based training](https://torchio.readthedocs.io/patches/index.html).

**Reply by gkrisp98** [reply]
Hi, again. 
I wanted to ask something about the training loop. From the tutorials I got this code:
```
for i, traindata in enumerate(train_loader):
        images = F.interpolate(traindata['t1'][tio.DATA], scale_factor=(1,1,1)).to('cuda:0')
        labels = traindata['label'].to('cuda:0')
```

I am not quite sure that I understand what the scale factor is, wich values I should assign and what shape should it have.
The train data has shape [1,224,224,260]. In the training loop I run I use it as (1,1,1) but when I try to run this line:
`image = F.interpolate(subjects_dataset[284]['t1'][tio.DATA], scale_factor=(1,1,1))`
for one instance of the dataset I get the error:
**scale_factor shape must match input shape. Input is 2D, scale_factor is 3**

Where have I mess up ?

**Reply by fepegar** [reply]
In which tutorial did you find that code?

**Reply by gkrisp98** [reply]
Oh yes I am sorry, it was from another user's repository. Is there a more efficient way to code this for the training and validation loop ?

**Reply by fepegar** [reply]
I don't know why the user wanted to do that, I think that does nothing. 

> Is there a more efficient way to code this

What would you like to code? I'm not sure what you mean.

**Reply by gkrisp98** [reply]
How to iterate through the dataloaders. To get images and labels, to pass them to model etc

**Reply by gkrisp98** [reply]
Update: I think I found it! Thanks a lot.

**Reply by fepegar** [reply]
Have you gone through the [tutorials](https://torchio.readthedocs.io/quickstart.html#hello-world)?

**Reply by gkrisp98** [reply]
I will check them again.

**Reply by gkrisp98** [reply]
Is there any tutorial for classification ?

**Reply by fepegar** [reply]
I don't think so. But classification is just a particular case of segmentation, so the segmentation tutorials should suffice.
